Index: gtk+3.0-3.14.13/gtk/gtkfilechooser.c
===================================================================
--- gtk+3.0-3.14.13.orig/gtk/gtkfilechooser.c
+++ gtk+3.0-3.14.13/gtk/gtkfilechooser.c
@@ -1433,7 +1433,7 @@ _gtk_file_chooser_log_event_finish (_Zei
     zeitgeist_event_add_subject (event, zeitgeist_subject_new_full (fdata->uri,
                                       NULL,
                                       NULL,
-                                      fdata->mime_type,
+                                      NULL,
                                       fdata->origin,
                                       fdata->display_name,
                                       NULL));
@@ -1486,6 +1486,11 @@ _gtk_file_chooser_log_event_finish (_Zei
 
   g_slist_free (data->fileinfos); // content freed in loop above
   g_slist_free (data->files); // content freed by g_object_unref (file) in the callback
+
+  // Throw the reference away, else we'll never be able to finalise chooser
+  g_object_unref (data->chooser);
+  data->chooser = NULL;
+
   g_free (data);
 }
 
@@ -1575,10 +1580,6 @@ _log_zeitgeist_event_add_file (_Zeitgeis
     #endif
 
     _gtk_file_chooser_log_event_finish (data, window_id);
-
-    // Throw the reference away, else we'll never be able to finalise chooser
-    g_object_unref (data->chooser);
-    data->chooser = NULL;
   }
 }
 
Index: gtk+3.0-3.14.13/gtk/gtkrecentchooser.c
===================================================================
--- gtk+3.0-3.14.13.orig/gtk/gtkrecentchooser.c
+++ gtk+3.0-3.14.13/gtk/gtkrecentchooser.c
@@ -785,7 +785,7 @@ _log_zeitgeist_get_recent_file_cb (GObje
   zeitgeist_event_add_subject (event, zeitgeist_subject_new_full (uri,
                                       NULL,
                                       NULL,
-                                      mime_type,
+                                      NULL,
                                       origin,
                                       display_name,
                                       NULL));
